// tsparticles-updater-orbit/umd/Options/Classes/Orbit.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tsparticles-engine", "./OrbitRotation"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Orbit = void 0;
    const tsparticles_engine_1 = require("tsparticles-engine");
    const OrbitRotation_1 = require("./OrbitRotation");
    class Orbit {
        constructor() {
            this.animation = new tsparticles_engine_1.AnimationOptions();
            this.enable = false;
            this.opacity = 1;
            this.rotation = new OrbitRotation_1.OrbitRotation();
            this.width = 1;
        }
        load(data) {
            if (!data) {
                return;
            }
            this.animation.load(data.animation);
            this.rotation.load(data.rotation);
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.opacity !== undefined) {
                this.opacity = (0, tsparticles_engine_1.setRangeValue)(data.opacity);
            }
            if (data.width !== undefined) {
                this.width = (0, tsparticles_engine_1.setRangeValue)(data.width);
            }
            if (data.radius !== undefined) {
                this.radius = (0, tsparticles_engine_1.setRangeValue)(data.radius);
            }
            if (data.color !== undefined) {
                this.color = tsparticles_engine_1.OptionsColor.create(this.color, data.color);
            }
        }
    }
    exports.Orbit = Orbit;
});
