// tsparticles-plugin-export-json/umd/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./ExportJSONInstance"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.loadExportJSONPlugin = void 0;
    const ExportJSONInstance_1 = require("./ExportJSONInstance");
    class ExportJSONPlugin {
        constructor(engine) {
            this.id = "export-json";
            this._engine = engine;
        }
        getPlugin(container) {
            return new ExportJSONInstance_1.ExportJSONInstance(container, this._engine);
        }
        loadOptions() {
        }
        needsPlugin() {
            return true;
        }
    }
    async function loadExportJSONPlugin(engine, refresh = true) {
        await engine.addPlugin(new ExportJSONPlugin(engine), refresh);
    }
    exports.loadExportJSONPlugin = loadExportJSONPlugin;
});
