// tsparticles-updater-orbit/umd/OrbitUpdater.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tsparticles-engine", "./Options/Classes/Orbit", "./Utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.OrbitUpdater = void 0;
    const tsparticles_engine_1 = require("tsparticles-engine");
    const Orbit_1 = require("./Options/Classes/Orbit");
    const Utils_1 = require("./Utils");
    class OrbitUpdater {
        constructor(container) {
            this.container = container;
        }
        afterDraw(particle) {
            const orbitOptions = particle.options.orbit;
            if (orbitOptions?.enable) {
                this.drawOrbit(particle, "front");
            }
        }
        beforeDraw(particle) {
            const orbitOptions = particle.options.orbit;
            if (orbitOptions?.enable) {
                this.drawOrbit(particle, "back");
            }
        }
        drawOrbit(particle, type) {
            const container = this.container;
            let start, end;
            switch (type) {
                case "back":
                    start = Math.PI / 2;
                    end = (Math.PI * 3) / 2;
                    break;
                case "front":
                    start = (Math.PI * 3) / 2;
                    end = Math.PI / 2;
                    break;
                default:
                    start = 0;
                    end = 2 * Math.PI;
            }
            container.canvas.draw((ctx) => {
                (0, Utils_1.drawEllipse)(ctx, particle, particle.orbitColor ?? particle.getFillColor(), particle.retina.orbitRadius ?? container.retina.orbitRadius ?? particle.getRadius(), particle.orbitOpacity ?? 1, particle.orbitWidth ?? 1, (particle.orbitRotation ?? 0) * container.retina.pixelRatio, start, end);
            });
        }
        init(particle) {
            const container = this.container, particlesOptions = particle.options, orbitOptions = particlesOptions.orbit;
            if (!orbitOptions?.enable) {
                return;
            }
            particle.orbitRotation = (0, tsparticles_engine_1.getRangeValue)(orbitOptions.rotation.value);
            particle.orbitColor = (0, tsparticles_engine_1.rangeColorToHsl)(orbitOptions.color);
            particle.retina.orbitRadius =
                orbitOptions.radius !== undefined
                    ? (0, tsparticles_engine_1.getRangeValue)(orbitOptions.radius) * container.retina.pixelRatio
                    : undefined;
            container.retina.orbitRadius = particle.retina.orbitRadius;
            particle.orbitAnimationSpeed = orbitOptions.animation.enable ? (0, tsparticles_engine_1.getRangeValue)(orbitOptions.animation.speed) : 0;
            particle.orbitWidth = (0, tsparticles_engine_1.getRangeValue)(orbitOptions.width);
            particle.orbitOpacity = (0, tsparticles_engine_1.getRangeValue)(orbitOptions.opacity);
        }
        isEnabled(particle) {
            const orbitAnimations = particle.options.orbit?.animation;
            return !particle.destroyed && !particle.spawning && !!orbitAnimations?.enable;
        }
        loadOptions(options, ...sources) {
            if (!options.orbit) {
                options.orbit = new Orbit_1.Orbit();
            }
            for (const source of sources) {
                options.orbit.load(source?.orbit);
            }
        }
        update(particle, delta) {
            if (!this.isEnabled(particle)) {
                return;
            }
            if (particle.orbitRotation === undefined) {
                particle.orbitRotation = 0;
            }
            particle.orbitRotation += (particle.orbitAnimationSpeed ?? 0 / (Math.PI * 2)) * delta.factor;
        }
    }
    exports.OrbitUpdater = OrbitUpdater;
});
