// tsparticles-plugin-polygon-mask/esm/index.js
import "./pathseg";
import { PolygonMask } from "./Options/Classes/PolygonMask";
import { PolygonMaskInstance } from "./PolygonMaskInstance";
class PolygonMaskPlugin {
    constructor(engine) {
        this.id = "polygonMask";
        this._engine = engine;
    }
    getPlugin(container) {
        return new PolygonMaskInstance(container, this._engine);
    }
    loadOptions(options, source) {
        if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
            return;
        }
        let polygonOptions = options.polygon;
        if (polygonOptions?.load === undefined) {
            options.polygon = polygonOptions = new PolygonMask();
        }
        polygonOptions.load(source?.polygon);
    }
    needsPlugin(options) {
        return (options?.polygon?.enable ??
            (options?.polygon?.type !== undefined && options.polygon.type !== "none"));
    }
}
export async function loadPolygonMaskPlugin(engine, refresh = true) {
    await engine.addPlugin(new PolygonMaskPlugin(engine), refresh);
}
export * from "./Enums/PolygonMaskInlineArrangement";
export * from "./Enums/PolygonMaskMoveType";
export * from "./Enums/PolygonMaskType";
