// tsparticles-plugin-polygon-mask/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPolygonMaskPlugin = void 0;
require("./pathseg");
const PolygonMask_1 = require("./Options/Classes/PolygonMask");
const PolygonMaskInstance_1 = require("./PolygonMaskInstance");
class PolygonMaskPlugin {
    constructor(engine) {
        this.id = "polygonMask";
        this._engine = engine;
    }
    getPlugin(container) {
        return new PolygonMaskInstance_1.PolygonMaskInstance(container, this._engine);
    }
    loadOptions(options, source) {
        if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
            return;
        }
        let polygonOptions = options.polygon;
        if (polygonOptions?.load === undefined) {
            options.polygon = polygonOptions = new PolygonMask_1.PolygonMask();
        }
        polygonOptions.load(source?.polygon);
    }
    needsPlugin(options) {
        return (options?.polygon?.enable ??
            (options?.polygon?.type !== undefined && options.polygon.type !== "none"));
    }
}
async function loadPolygonMaskPlugin(engine, refresh = true) {
    await engine.addPlugin(new PolygonMaskPlugin(engine), refresh);
}
exports.loadPolygonMaskPlugin = loadPolygonMaskPlugin;
__exportStar(require("./Enums/PolygonMaskInlineArrangement"), exports);
__exportStar(require("./Enums/PolygonMaskMoveType"), exports);
__exportStar(require("./Enums/PolygonMaskType"), exports);
