// tsparticles-plugin-export-image/umd/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./ExportImageInstance"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.loadExportImagePlugin = void 0;
    const ExportImageInstance_1 = require("./ExportImageInstance");
    class ExportImagePlugin {
        constructor(engine) {
            this.id = "export-image";
            this._engine = engine;
        }
        getPlugin(container) {
            return new ExportImageInstance_1.ExportImageInstance(container, this._engine);
        }
        loadOptions() {
        }
        needsPlugin() {
            return true;
        }
    }
    async function loadExportImagePlugin(engine, refresh = true) {
        await engine.addPlugin(new ExportImagePlugin(engine), refresh);
    }
    exports.loadExportImagePlugin = loadExportImagePlugin;
});
