// tsparticles-plugin-export-video/umd/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./ExportVideoInstance"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.loadExportVideoPlugin = void 0;
    const ExportVideoInstance_1 = require("./ExportVideoInstance");
    class ExportVideoPlugin {
        constructor(engine) {
            this.id = "export-video";
            this._engine = engine;
        }
        getPlugin(container) {
            return new ExportVideoInstance_1.ExportVideoInstance(container, this._engine);
        }
        loadOptions() {
        }
        needsPlugin() {
            return true;
        }
    }
    async function loadExportVideoPlugin(engine, refresh = true) {
        await engine.addPlugin(new ExportVideoPlugin(engine), refresh);
    }
    exports.loadExportVideoPlugin = loadExportVideoPlugin;
});
